src/histogram.c

```diff
@@ -29,12 +38,226 @@ TS_FUNCTION_INFO_V1(hist_serializefunc);
+} HistogramSerializeState;
+
+static inline void
+set_first_tuples(HistogramState *state, bool val)
+{
+	Size		i;
+	Size		arrsize = HIST_LEN(state->values);
+
+	for (i = 0; i < arrsize; i++)
+		state->first_tuple[i] = val;
+}
+
+/* Copied from postgres source */
+static Datum
+get_agg_init_val(Datum textInitVal, Oid transtype, MemoryContext ctxt)
+{
+	Oid			typinput;
+	Oid			typioparam;
+	char	   *strInitVal;
+	Datum		initVal;
+	MemoryContext old = MemoryContextSwitchTo(ctxt);
+
+	getTypeInputInfo(transtype, &typinput, &typioparam);
+	strInitVal = TextDatumGetCString(textInitVal);
+	initVal = OidInputFunctionCall(typinput, strInitVal,
+								   typioparam, -1);
+	pfree(strInitVal);
+	MemoryContextSwitchTo(old);
+	return initVal;
+}
+
+static inline HistogramState *
+init_state(MemoryContext aggcontext, Size arrsize)
+{
+	HistogramState *state = (HistogramState *) MemoryContextAlloc(aggcontext, sizeof(HistogramState));
+
+	state->values = MemoryContextAllocZero(aggcontext, VALUES_SIZE(arrsize));
+	state->first_tuple = MemoryContextAlloc(aggcontext, arrsize);
+	SET_VARSIZE(state->values, VALUES_SIZE(arrsize));
+	set_first_tuples(state, true);
+	state->subagg_state = (SubAggState)
+	{
+		.sfunc = InvalidOid,
+			.combinefunc = InvalidOid,
+			.finalfunc = InvalidOid,
+			.serialfunc = InvalidOid,
+			.deserialfunc = InvalidOid,
+			.result_type = InvalidOid,
+			.isstrict_sfunc = false,
+	};
+
+	return state;
+}
+
+static Oid
+find_available_aggregate(char *subagg_text)
+{
+	List	   *agg_fn_name = list_make1(makeString(subagg_text));
+	FuncCandidateList funclist = FuncnameGetCandidates(agg_fn_name, 1, NULL,
+													   false, true, false);
+
+	/*
+	 * Currently we only support histogram on FLOAT8 subtype columns, so check
+	 * if we can find a NUMERIC or ANY version of the aggregator
+	 */
+	while (funclist != NULL && (funclist->nargs != 1 ||
+								(funclist->args[0] != FLOAT8OID && funclist->args[0] != ANYOID)))
+	{
+		funclist = funclist->next;
+	}
+
+	if (NULL == funclist)
+		elog(ERROR, "Aggregation function not found");
+
+	return funclist->oid;
+}
+
+static void
+set_bucket_initial_values(HistogramInternalState *istate, HistogramState *state,
+						  MemoryContext ctxt)
+{
+	if (istate->init_value_isnull)
+		set_first_tuples(state, true);
+	else
+	{
+		Size		i;
+		Size		arrsize = HIST_LEN(state->values);
+		Datum	   *hist = (Datum *) VARDATA(state->values);
+
+		for (i = 0; i < arrsize; i++)
+			hist[i] = get_agg_init_val(istate->text_init_value,
+									   istate->trans_type, ctxt);
+		set_first_tuples(state, false);
+	}
+}
+
+static Datum
+apply_subagg_sfunc(Oid funcid, fmNodePtr ctxt, Datum arg1, Datum arg2)
+{
+	FmgrInfo	flinfo;
+	FunctionCallInfoData fcinfo;
+	Datum		res;
+
+	fmgr_info(funcid, &flinfo);
+	InitFunctionCallInfoData(fcinfo, &flinfo, 2, InvalidOid, ctxt, NULL);
+	fcinfo.arg[0] = arg1;
+	fcinfo.arg[1] = arg2;
+	fcinfo.argnull[0] = false;
+	fcinfo.argnull[1] = false;
+
+	res = FunctionCallInvoke(&fcinfo);
```

Comment:
Any reason we are not using `OidFunctionCall2()` here? I think that would be a lot cleaner if possible. Is it because you need to pass on the context via `ctxt`?

Reply:
Need to pass `ctxt` so unfortunately cannot (same below)